### ISSUE TITLE
Issue with inference using fine-trained model

### ISSUE THREAD

**Issue by rhobincu** [msminhas93/DeepLabv3FineTuning#4]
After retraining the model using the code in this repository, I have attempted to use it in order to segment one of the images in the CrackForest db. The code is taken from the pytorch page:
```python
import torch
model = torch.load('output/weights.pt')
model.eval()

import urllib
url, filename = ("file:///home/rhobincu/gitroot/DeepLabv3FineTuning/CrackForest/Images/092.jpg", "092.jpg")
try: urllib.URLopener().retrieve(url, filename)
except: urllib.request.urlretrieve(url, filename)

# sample execution (requires torchvision)
from PIL import Image
from torchvision import transforms
input_image = Image.open(filename)
print(input_image)
preprocess = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
])

input_tensor = preprocess(input_image)
input_batch = input_tensor.unsqueeze(0) # create a mini-batch as expected by the model

# move the input and model to GPU for speed if available
if torch.cuda.is_available():
    print('Using GPU!')
    input_batch = input_batch.to('cuda')
    model.to('cuda')

import time

start = time.clock()
with torch.no_grad():
    output = model(input_batch)['out'][0]
end = time.clock()
print('Inference duration (s): ', end - start)
output_predictions = output.argmax(0)

# create a color pallette, selecting a color for each class
palette = torch.tensor([2 ** 25 - 1, 2 ** 15 - 1, 2])
colors = torch.as_tensor([i for i in range(2)])[:, None] * palette
colors = (colors % 255).numpy().astype("uint8")
print(colors)

# plot the semantic segmentation predictions of 21 classes in each color
img_size = input_image.size
data = output_predictions.byte().cpu().numpy()
print(data)
print(data.sum())
r = Image.fromarray(data).resize(img_size)
r.putpalette(colors)

#cv2.imshow('image',input_image)

import matplotlib.pyplot as plt
plt.imshow(r)
plt.show()
```

The problem is that the output from the nn (data) is full of 0. Any ideas?

**Comment by msminhas93** [comment]
Don't do the normalization. Just divide by 255 if you are using my model.

**Comment by rhobincu** [comment]
Hello, thank you for your time! I have tried that too (the same result - all 0 output), with the following code:
```python
preprocess = transforms.Compose([
    transforms.ToTensor(),
])

input_tensor = preprocess(input_image)
input_tensor = input_tensor.type(torch.FloatTensor) / 255.0
```

This is the output of the training log:
```
epoch,Train_loss,Test_loss,Train_f1_score,Train_auroc,Test_f1_score,Test_auroc
1,0.028723619878292084,0.39652174711227417,0.07258603287886232,0.608278830132879,0.015250583915610242,0.4454309197809733
2,0.01655764877796173,0.12919872999191284,0.1764006943081475,0.8207898913057353,0.14374211618803936,0.7250351687571401
3,0.013691616244614124,0.015931623056530952,0.2528306810489227,0.8867562918617247,0.2665119007076772,0.7982250048754536
4,0.009655261412262917,0.018562788143754005,0.30563098594493987,0.9109607614186882,0.25672394252339414,0.77547558097446
5,0.013742960058152676,0.00749854976311326,0.34047307078284184,0.9174267806289312,0.298076998358514,0.7653140712243138
6,0.014444611966609955,0.016985420137643814,0.3514685032038002,0.9261002980566492,0.3390751560813173,0.7731331231859201
7,0.010918714106082916,0.3412984311580658,0.3557727693681669,0.9236153221253032,0.19926198145184562,0.7619734968787623
8,0.009675263427197933,0.006430158391594887,0.3654654584004751,0.9264135928406162,0.324659376316098,0.8188795900443927
9,0.008248267695307732,0.008450563065707684,0.3718889143997008,0.9288937903130362,0.3397139368577577,0.8207357873440143
10,0.00878356397151947,0.007887532003223896,0.3892080990929599,0.9312094778831088,0.2463976175030604,0.7544823510596433
11,0.011835074983537197,0.007407198194414377,0.38084296547851315,0.9298519117349485,0.3210037960339835,0.8032530907449412
12,0.011028565466403961,0.007152864709496498,0.3898585757468897,0.9346812544250851,0.3265875587725663,0.8028356081918123
13,0.009070910513401031,0.00602466706186533,0.39329011892366467,0.932107458863866,0.3417528596322977,0.796751455440501
14,0.01122941542416811,0.007177171763032675,0.38973680034693864,0.9310570299818463,0.35646529822567413,0.8044541916791449
15,0.008179922588169575,0.007051416672766209,0.3963375237657409,0.9350139244739992,0.32817196727185083,0.8130408750822863
16,0.024151064455509186,0.007630373816937208,0.39718840358498225,0.935834002788122,0.3363901601681642,0.8062185288559528
17,0.011229011230170727,0.006691098678857088,0.40275023696556006,0.9365604049241338,0.34555403880441793,0.7999246162011975
18,0.010117234662175179,0.008310572244226933,0.4185488615227668,0.9358812561373238,0.3138476950418036,0.8092566566013656
19,0.01143378671258688,0.0069137937389314175,0.4041738484809898,0.9333158347590814,0.33091260407768397,0.8177593114010504
20,0.013196403160691261,0.00892605260014534,0.4197683783305407,0.9378494515419616,0.33172089214575406,0.7961911676603153
21,0.011065822094678879,0.007611274253576994,0.3954620372733617,0.9361832287014848,0.29645035286395094,0.8236634918867152
22,0.008103985339403152,0.013817863538861275,0.39031057651796375,0.9352408831296549,0.35468890921264373,0.8244557147687476
23,0.008143968880176544,0.012763076461851597,0.40727482883861926,0.9356572150767043,0.3343516517975274,0.8297717826912497
24,0.008757246658205986,0.00884932093322277,0.40482454768387915,0.9365414627179064,0.3173204570613852,0.7704947430593242
25,0.007630358915776014,0.007393880747258663,0.41739505236048724,0.9372066906851276,0.34069494149820906,0.8359078956632977
``` 

I am using python3 and the tensorflow-gpu package.

**Comment by rhobincu** [comment]
I'm adding the output of running the inference script:

```
rhobincu@daneel:~/gitroot/DeepLabv3FineTuning$ ./test.py 
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=480x320 at 0x7FB608938A58>
Using GPU!
Inference duration (s):  0.048487999999999865
[[  0   0   0]
 [  1 127   2]]
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
0
```

**Comment by msminhas93** [comment]
I'm sorry for the delayed response.

Did you check `data = output_predictions.byte().cpu().numpy()` matrix?   Since you are rounding off the values using `data = output_predictions.byte().cpu().numpy()`, I think all of the values got quantized to 0 because data values are less than 1.

Also, when you are using ToTensor() please do not divide by 255 after that since that will result in division by (255)**2.

**Comment by istomoya** [comment]
Size() of tensor "output" in
  output_predictions = output.argmax(0)
differs between using trained model and using the original deeplabv3-resnet101.
Using the trained model
output.size() is [1, 320, 480]  in my case. Here first dimension is "1".
Therefore argmax(0) always gives zero.
Using original deeplab-resnet101,
output.size() is [21, 332, 449]
so argmax(0) gives meaningful integer.

**Comment by adnankarim** [comment]
Is there any solution to this? I am getting exact same problem? Just black output...

**Comment by msminhas93** [comment]
> Is there any solution to this? I am getting exact same problem? Just black output...

Are you using the above script as well? The colour palette won't work the way the code uses since I'm using only a single channel for both classes. Did you try the notebook code in the repo for inference?

**Comment by adnankarim** [comment]
Thank you..That is working..

**Comment by adnankarim** [comment]
Can you suggest any changes to use 2 channels for classes. to the current so that It can be deployed to android via boiler plate code of 21 classes without any changes.
